### PR TITLE
Main performance curve

### PR DESCRIPTION
**Pull request by trant22t**

### CODE REVIEW

File: performance_curves/performance_curve.py [hunk 1]
```diff
@@ -1,35 +1,134 @@
...
 
 from performance_curves.utils import synchronize_sort
 
 
-class PerformanceCurve:
+class PerformanceCurveLike:
     """Stores values of the x- and y-axis of the performance curve and allows plotting"""
-    def __init__(self, cutoff, metric_value):
-        self.cutoff = cutoff
+    def __init__(self, metric_value, example_counts):
```
Comment (erp12) [inline performance_curves/performance_curve.py]: Some nit-picks on naming.

We should decide if we are using singular or plural for the array variables.

Also, we should select term for the following concepts and use that throughout the package:
- metric, performance, measurement, evaluation, etc.
- observation, case, example, etc.
- score, prediction, ranking, etc.

File: performance_curves/performance_curve.py [hunk 2]
```diff
@@ -1,35 +1,134 @@
...
+            performance_point: float,
+            comparison_operator: str,
+            count_upper_bound: Optional[int] = None,
+            count_lower_bound: Optional[int] = None,
+            cutoff_upper_bound: Optional[float] = None,
+            cutoff_lower_bound: Optional[float] = None,
+            highest: bool = False
+    ) -> float:
+        """
+        Returns score cutoff which meets a certain performance criteria given other optional counts and scores
```
Comment (erp12) [inline performance_curves/performance_curve.py]: Some more terms to standardize:
- Cutoff, threshold, decision boundary, etc.
- performance criteria, performance objective, performance constraint, etc.

File: performance_curves/performance_curve.py [hunk 3]
```diff
@@ -1,35 +1,134 @@
...
+            `cutoff_at(0.80, '>=', count_lower_bound=100)` returns the lowest score cutoff that produces a performance
+            of at least 0.80 and uses at least 100 observations.
+            `cutoff_at(0.75, '<', cutoff_upper_bound=0.25, highest=True)` returns the highest score cutoff less than
+            0.25 that produces a performance less than 0.75.
+        """
+
+        assert len(self.metric_value) == len(self.example_counts), \
+            'Lengths of performance values array and counts array do not match.'
+        assert len(self.metric_value) == len(self.score_cutoff), \
+            'Lengths of performance values array and probability estimates array do not match.'
```
Comment (erp12) [inline performance_curves/performance_curve.py]: We can perform these checks once in the constructor.

File: performance_curves/performance_curve.py [hunk 4]
```diff
@@ -1,35 +1,134 @@
...
+        ops = {'>': operator.gt,
+               '<': operator.lt,
+               '>=': operator.ge,
+               '<=': operator.le,
+               '==': operator.eq}
+        if comparison_operator in ops:
+            metric_ind = np.where(ops[comparison_operator](self.metric_value, performance_point))[0]
+        else:
+            raise ValueError("comparison_operator argument could only take one of the following strings: "
+                             "'>', '<', '>=', '<=', '=='.")
```
Comment (erp12) [inline performance_curves/performance_curve.py]: It looks like `comparison_operator` is logically meant to map to a function (from the operator module). Are we gaining anything by having the user pass in a string argument? Why not pass in the function as the argument?

File: performance_curves/performance_curve.py [hunk 5]
```diff
@@ -1,35 +1,134 @@
...
+
+        bound_ind = np.array([i for i in range(len(self.metric_value))])
```
Comment (erp12) [inline performance_curves/performance_curve.py]: Faster and more readable as `np.arange()`

File: performance_curves/performance_curve.py [hunk 6]
```diff
@@ -1,35 +1,134 @@
...
+    def cutoff_at(
+            self,
+            performance_point: float,
+            comparison_operator: str,
+            count_upper_bound: Optional[int] = None,
+            count_lower_bound: Optional[int] = None,
+            cutoff_upper_bound: Optional[float] = None,
+            cutoff_lower_bound: Optional[float] = None,
+            highest: bool = False
+    ) -> float:
```
Comment (erp12) [inline performance_curves/performance_curve.py]: Interesting redesign into a single function. Some questions and things to think about:

- Does changing the comparison operator meaningfully change the result? In my mental model, we are trying to identify score cutoffs where the model's performance crosses the desired performance objective. 

- I can see how using a single function with the `highest` argument might DRY out the code a bit. There might be other ways to accomplish the same thing that maintain function names with clear semantics.

File: performance_curves/performance_curve.py [hunk 7]
```diff
@@ -1,35 +1,134 @@
...
+        if cutoff_lower_bound is not None:
+            cutoff_lower_bound_ind = np.where(self.score_cutoff >= cutoff_lower_bound)[0]
+            bound_ind = np.intersect1d(bound_ind, cutoff_lower_bound_ind)
+
+        result_ind = np.intersect1d(metric_ind, bound_ind)
+        if len(result_ind) > 0:
+            if highest:
+                return self.score_cutoff[result_ind[0]]
+            else:
+                return self.score_cutoff[result_ind[-1]]
```
Comment (erp12) [inline performance_curves/performance_curve.py]: Suppose we ask for the cutoff with the `performance_point` of 0.85 with the `==` operator and the following performance curve:

| score cutoffs | metric values |
|---------------|---------------|
| 0.9           | 0.95          |
| 0.8           | 0.87          |
| 0.7           | 0.77          |

It seems to me this code would return `None` when it is clear that this performance curve does cross a performance of 0.85 at some point. We should be able to make a "best effort" answer at a score cutoff in this situation. I'm not 100% sure what the best way to do this is, but I have 2 potential approaches.

1. Return the score nearest cutoff where performance still satisfies the metric goal. (Precision and recall are both metrics we maximize, so pick we would pick the cutoff that is closest to but still above the goal. For metrics which we minimize, we would pick the closest that is below the goal)

2. Use linear interpolation to make an estimated guess at the exact score cutoff.

**Review by erp12 — CHANGES_REQUESTED** [review]
Let me know what your preferences are on standardizing terminology.

File: performance_curves/performance_curve.py (hunk 1 above)
Comment (trant22t) [reply performance_curves/performance_curve.py]: Great call! 

For array variables, I have changed them all to plural (e.g. `performance_values` or `case_counts`) since the array contains multiple items. 

For consistent terminology, I now use the terms as follows:
- metric, performance, etc --> performance
- observation, case, example, etc. --> case
- score, prediction, probability estimate --> score
- cutoff, threshold, etc --> threshold

I have updated these changes in this PR, but let me know if you think differently.

File: performance_curves/performance_curve.py (hunk 2 above)
Comment (trant22t) [reply performance_curves/performance_curve.py]: Updated! 

- cutoff, threshold, etc. --> cutoff
- criterion (criteria for plural), constraint --> criterion/criteria

File: performance_curves/performance_curve.py (hunk 5 above)
Comment (trant22t) [reply performance_curves/performance_curve.py]: Another good call! Updated. Thank you!

File: performance_curves/performance_curve.py (hunk 3 above)
Comment (trant22t) [reply performance_curves/performance_curve.py]: I moved these assertions to the end of `performance_curve` function.

Comment (trant22t) [reply performance_curves/performance_curve.py]: Just kidding. Moved them inside the constructor

File: performance_curves/performance_curve.py [hunk 8]
```diff
@@ -1,78 +1,123 @@
...
+                 performance_values: np.ndarray,
+                 case_counts: np.ndarray,
+                 metric: Metric):
+        self.performance_values = performance_values
+        self.case_counts = case_counts
+        self.metric = metric
+
+    def plot(self, target_performance_values: Optional[List[float]] = None):
+        if target_performance_values is not None and not isinstance(self, PerformanceCurve):
+            raise ValueError("Cannot annotate target performance values of " + type(self).__name__)
```
Comment (erp12) [inline performance_curves/performance_curve.py:22]: Let's test that a `ValueError` is raised when passed Random/Perfect curves.

File: performance_curves/performance_curve.py [hunk 9]
```diff
@@ -1,78 +1,123 @@
...
+class PerformanceCurve(NonRandomPerformanceCurve):
+    def __init__(self,
+                 y_true: np.ndarray,
+                 y_score: np.ndarray,
+                 metric: Metric,
+                 num_bins: Optional[int] = None,
+                 order_descending: bool = True):
+        self.y_true = y_true
+        self.y_score = y_score
+        assert len(self.y_score) == len(self.y_true), "Lengths of ground-truth and prediction arrays do not match."
```
Comment (erp12) [inline performance_curves/performance_curve.py:55]: Assertion needs test.

File: tests/test_utils.py [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+
+def test_synchronize_sort():
+    key = np.array([0.1, 0.4, 0.91, 0.8, 0.05])
+    dependent = np.array([0, 0, 1, 1, 0])
+
+    sorted_key, sorted_dependent = synchronize_sort(key, dependent, descending=True)
+
+    np.testing.assert_array_equal(sorted_key, np.array([0.91, 0.8, 0.4, 0.1, 0.05]))
+    np.testing.assert_array_equal(sorted_dependent, np.array([1, 1, 0, 0, 0]))
```
Comment (erp12) [inline tests/test_utils.py]: Some other edge cases to test:
- What should happen when the arrays are different sizes?
- Given empty arrays.

File: tests/test_utils.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+
+def test_get_bin_sizes():
+    scores = np.array([0.1, 0.4, 0.91, 0.8, 0.05])
+    bin_sizes = get_bin_sizes(scores, 3)
+    np.testing.assert_array_equal(bin_sizes, [2, 2, 1])
```
Comment (erp12) [inline tests/test_utils.py]: Some additional edge cases:
- Empty array
- Zero bins
- Bin count higher than the length of the array.

**Review by erp12 — CHANGES_REQUESTED** [review]
Design and logic looks good. I just have some minor suggestions for additional tests.

File: tests/test_utils.py (hunk 11 above)
Comment (trant22t) [reply tests/test_utils.py]: All very good edge cases to test! Thank you for catching them. I have updated these tests as suggested.

File: tests/test_performance_curve.py [hunk 12]
```diff
@@ -1,62 +1,112 @@
...
-    np.testing.assert_array_equal(x2.cutoff, sorted_y_true_ub)
-    np.testing.assert_array_equal(x3.cutoff, sorted_y_true_ub)
-    np.testing.assert_array_equal(x4.cutoff, sorted_y_true_ub)
-
-    np.testing.assert_array_almost_equal(x1.metric_value, np.array([1., 0.5, 0.4]), decimal=3)
-    np.testing.assert_array_almost_equal(x2.metric_value, np.array([1., 1., 1.]), decimal=3)
-    np.testing.assert_array_almost_equal(x3.metric_value, np.array([1., 0.6, 0.4]), decimal=3)
-    np.testing.assert_array_almost_equal(x4.metric_value, np.array([1., 0.666, 0.5]), decimal=3)
+    # Expect to raise error
+    with np.testing.assert_raises(AssertionError):
```
Comment (erp12) [inline tests/test_performance_curve.py:69]: Interesting choice to use numpy testing utilities for this. I didn't know they provided this functionality. No need to make a change, but I'm curious if there is a reason to use this instead of [the same functionality provided by `pytest`](https://docs.pytest.org/en/stable/assert.html#assertions-about-expected-exceptions). 

```python
with pytest.raises(AssertionError):
    ...
```

**Review by erp12 — APPROVED** [review]
Looks great!

File: tests/test_performance_curve.py (hunk 12 above)
Comment (trant22t) [reply tests/test_performance_curve.py:69]: The main reason is just because my PyCharm keeps complaining that I need to include `pytest` in the project requirements and I think we agreed that `requirements.txt` should only include necessary packages for the users (not developers). I know that we already included `pytest` in `tox.ini`, but not sure why my PyCharm still complains. Anyway, since we already use `np.testing` for all the test cases above and there's not any significant difference between the two ways of raising exceptions, I figure it would be safe to use `np.testing.assert_raises()` in this case.
